[flang][OpenMP] Fix copyprivate semantic checks

luporl opened the pull request:
There are some cases in which variables used in OpenMP constructs
are predetermined as private. The semantic checks for copyprivate
were not handling those cases.

Besides that, shared symbols were not being properly represented
in some cases. When there was no previously declared private
(implicit) symbol, no new association symbols, representing
shared ones, were being created.

These symbols must always be inserted in constructs that may
privatize the original symbol: parallel, teams and task
generating constructs.

Fixes #87214 and #86907

luporl committed:
[flang][OpenMP] Fix copyprivate semantic checks

There are some cases in which variables used in OpenMP constructs
are predetermined as private. The semantic checks for copyprivate
were not handling those cases.

Fixes #87214 and #86907

tblah commented on flang/lib/Semantics/resolve-directives.cpp:
```diff
@@ -2540,6 +2541,32 @@ void ResolveOmpTopLevelParts(
+  return false;
+}
+
+static bool IsSymbolPredeterminedAsPrivate(const Symbol &symbol) {
```
What about loop indices?

luporl replied:
Loop indices are handled by other parts of the OmpAttributeVisitor, that should run before this check and set the OmpPrivate flag in the corresponding symbols. However, I'm seeing some strange behaviors with the copyprivate check in some cases. I'll investigate it further and add some more tests.

luporl committed:
Fix handling of shared symbols and refactoring

Shared symbols were not being properly represented in some cases.
When there was no previously declared private symbol, no new
association symbols, representing shared ones, were being created.

These symbols should always be inserted in constructs that may
privatize the original symbol: parallel, teams and task generating
constructs.

Several tests had to be updated, to expect these previously missing
symbols.

luporl commented:
The copyprivate checks should behave correctly now, in all cases that I'm aware of.
Some symbols that represent variables with shared DSA were missing too. These were added, which required the update of some tests. I also did a slight code refactoring.

tblah reviewed (commented):
LGTM, but please wait for review from somebody else.

luporl commented:
Thanks for the review @tblah.

Ping for other reviews.

NimishMishra commented:
> Ping for other reviews.

Sorry for the delay. I'll a look and revert, latest by tomorrow.

NimishMishra commented on flang/lib/Semantics/resolve-directives.cpp:
```diff
@@ -2540,20 +2554,58 @@ void ResolveOmpTopLevelParts(
-    const parser::Name &name, const Symbol &symbol, Symbol::Flag ompFlag) {
-  const auto *checkSymbol{&symbol};
+static bool IsSymbolInCommonBlock(const Symbol &symbol) {
+  // If there are many symbols in common blocks, going through them all can be
```
Would you like this to be a TODO, instead of a comment, depending on how impactful we expect this to be?

luporl replied:
Thanks for the review!

Comment changed to a TODO.

For now I don't expect the current implementation of `IsSymbolInCommonBlock` to have a big impact on performance, since it's used only when checking symbols in `copyprivate` clauses. But on programs with an unusually high number of symbols in common blocks and `copyprivate` clauses, or if this predicate function starts to be used within other checks, then it could be a problem.

NimishMishra reviewed (approved):
Thanks for the patch. LGTM overall

luporl committed:
Address review's comments

psteinfeld commented:
@luporl, after this change, I'm seeing errors in our internal test suites.  Here's a program that produces similar errors:
```
[pete@rome3 install]$cat bug.f90
program bug
  use omp_lib
  implicit none
  integer i1,i2,i3,m
  type derived
    integer field(2,3)
  end type
  type(derived) y(5)
  integer,parameter :: nt = 2
  integer field(nt)
  field = 0
  m = 1
  !$OMP PARALLEL DO NUM_THREADS(nt) PRIVATE(m) DEFAULT(NONE) SHARED(y,field) &
  !$OMP COLLAPSE(2) SCHEDULE(STATIC)
  do i1=1,size(y,1)
    do i3=1,3
      do i2=1,2
        y(i1)%field(i2,i3) = 1
        !$omp atomic
        field(m) = field(m) + 1 !error here
      end do
    end do
  end do
  !$OMP END PARALLEL DO
end program

flang-new -fopenmp -fopenmp-version=31 bug.f90 
error: Semantic errors in bug.f90
./bug.f90:20:9: error: Reference to rank-2 object 'field' has 1 subscripts
          field(m) = field(m) + 1 !error here
          ^^^^^^^^
./bug.f90:20:20: error: Reference to rank-2 object 'field' has 1 subscripts
          field(m) = field(m) + 1 !error here
                     ^^^^^^^^
```

Before you merged this change, this program compiled cleanly.  Can you please either revert or fix this?

luporl commented:
@psteinfeld Reverted. I'll need some time to investigate this error.

tblah commented:
A minimised test of ours which failed too (posting in addition in case it helps the investigation):

```
program omp_associate_min
  integer, parameter :: l = 42
  integer :: a(l)
  integer :: i
  a = 1

  !$omp parallel do
  do i = 1,l
    associate (b=>a)
      b(i) = b(i) * 2
    end associate
  enddo
  !$omp end parallel do
end program
```

Thanks for the quick revert and cherry-pick to the release branch.

luporl commented:
Thanks for the feedback @psteinfeld and @tblah.
The regressions should be fixed in https://github.com/llvm/llvm-project/pull/101009.
Please check if it fixes your tests.
